### DISCUSSION TITLE
Cell elements automatically merge in pyvista.UnstructuredGrid.add_mesh_clip_box (bug or feature?)

### DISCUSSION THREAD

**Discussion by basnijholt** [pyvista/pyvista#2168] [Q&A] [answered]
I want to have independent mesh elements such that I am able to show discontinuities.

See the following function, where I create two separately defined triangles:
```python
import numpy as np
import pyvista as pv

def get_data(eps=1e-7):
    points = np.array(
        [
            [0, 0, 0],
            [1 - eps, 0, 0],
            [0, 1 - eps, 0],
            [1, 0, 0],
            [1, 1, 0],
            [0, 1, 0],
        ]
    )
    topology = np.array([[3, 0, 1, 2], [3, 3, 4, 5]])
    cell_types = np.full(2, 5)
    data = pv.UnstructuredGrid(topology, cell_types, points)
    data.point_data["f"] = np.array([0, 0, 0, 1, 1, 1])
    return data
```

I have the following behaviour:

```
data = get_data(eps=0) # eps=1e-7 is the same!!!!!!!
p = pv.Plotter()
p.add_mesh(data)
p.show()
```
![image](https://user-images.githubusercontent.com/6897215/153262883-c42755a5-239c-4689-a9d4-5ad938dd5601.png)

However, when using `add_mesh_clip_box`:

```
data = get_data(eps=0)
p = pv.Plotter()
p.add_mesh_clip_box(data, show_edges=True)
p.show()
```
![image](https://user-images.githubusercontent.com/6897215/153263061-589d68a8-86aa-4c6b-b8a8-bab9378e47e6.png)

Now when I move the corner points of the triangles by a small number I get what I initially expect (when not moving the triangles):
```
data = get_data(eps=1e-7)
p = pv.Plotter()
p.add_mesh_clip_box(data, show_edges=True)
p.show()
```
![image](https://user-images.githubusercontent.com/6897215/153263227-a33fad37-51de-4edd-9085-8673987e147e.png)

Why does `add_mesh_clip_box` merge the triangles?

**Comment by banesullivan** [comment]
> Why does add_mesh_clip_box merge the triangles?

I'm not sure I understand what you mean by this. Are you referring the color mapping differences?

**Reply by basnijholt** [reply]
The two points where the triangles intersect have different values. So at coordinate `(0, 1, 0)` the value for the first triangle is `0`, however, the value of the second triangle at coordinate `(0, 1, 0)` is `1`. That is the reason why I expect a discontinuity to show up, and not a gradient.

**Reply by banesullivan** [reply]
Ah, I did not realize they had different values/are different points.

See comment below

**Comment by basnijholt** [comment, marked as answer]
I think this might be related to the https://github.com/pyvista/pyvista/blob/06b1ac01214a4c636395f139b681acea2543960c/pyvista/plotting/widgets.py#L184

because this uses the [(VTK docs)](https://vtk.org/doc/nightly/html/classvtkBoxClipDataSet.html#a6d87d4d98a2f435badf276a076cc9218) `vtkMergePoints` Locator. I have no idea what this means though. Perhaps I need to call `alg.SetLocator(vtkPointLocator)` or something?

**Reply by banesullivan** [reply]
as you found, `add_mesh_clip_box` uses `vtkBoxClipDataSet ` under the hood:

https://github.com/pyvista/pyvista/blob/06b1ac01214a4c636395f139b681acea2543960c/pyvista/plotting/widgets.py#L184

This VTK filter will, by default, merge points that are sufficiently close. See https://github.com/Kitware/VTK/blob/01f5cc18fd9f0c8e34a5de313d53d2178ff6e325/Filters/General/vtkBoxClipDataSet.cxx#L522

What we could do is add a parameter to change the locator of this algorithm to disable merging the points as the default locator is `vtkMergePoints`. I think a vanilla `vtkPointLocator` will prevent this

**Reply by basnijholt** [reply]
Great! vtkPointLocator didn't work, but the following did:

```
from vtkmodules.vtkCommonDataModel import vtkNonMergingPointLocator
alg.SetLocator(vtkNonMergingPointLocator())
```

:tada:

**Reply by basnijholt** [reply]
I have opened a PR: https://github.com/pyvista/pyvista/pull/2172.
